src/docfx/resolve/Resolve.cs

```diff
@@ -9,53 +9,51 @@ namespace Microsoft.Docs.Build
 {
     internal static class Resolve
     {
-        public static (string content, Document file) TryResolveContent(this Document relativeTo, string href)
+        public static (DocfxException error, string content, Document file) TryResolveContent(this Document relativeTo, string href)
         {
-            var (file, _) = TryResolveFile(relativeTo, href);
+            var (error, file, _) = TryResolveFile(relativeTo, href);
 
-            return file != null ? (file.ReadText(), file) : default;
+            return file != null ? (error, file.ReadText(), file) : default;
         }
 
-        public static (string href, Document file) TryResolveHref(this Document relativeTo, string href, Document resultRelativeTo = null)
+        public static (DocfxException error, string href, Document file) TryResolveHref(this Document relativeTo, string href, Document resultRelativeTo)
         {
-            var (file, fragmentQuery) = TryResolveFile(relativeTo, href);
+            Debug.Assert(resultRelativeTo != null);
+
+            var (error, file, fragmentQuery) = TryResolveFile(relativeTo, href);
 
             // Cannot resolve the file, leave href as is
-            if (file == null)
+            if (file == null || file == relativeTo)
             {
-                return (href, null);
+                return (error, href, file);
             }
 
             var resolvedHref = file.SiteUrl + fragmentQuery;
 
-            if (resultRelativeTo == null)
-            {
-                return (resolvedHref, file);
-            }
-
             // Make result relative to `resultRelativeTo`
             resolvedHref = PathUtility.GetRelativePathToFile(resultRelativeTo.SiteUrl, file.SiteUrl).Replace('\\', '/');
 
-            return (resolvedHref + fragmentQuery, file);
+            return (error, resolvedHref + fragmentQuery, file);
         }
 
-        private static (Document file, string fragmentQuery) TryResolveFile(this Document relativeTo, string href)
+        private static (DocfxException error, Document file, string fragmentQuery) TryResolveFile(this Document relativeTo, string href)
```

Comment:
>DocfxException error [](start = 24, length = 20)

feels strange that we return, instead of throw an exception

Reply:
If you must catch immediately after the call, why not return it :)

---
In reply to: [193965833](https://github.com/dotnet/docfx/pull/2857#discussion_r193965833) [](ancestors = 193965833)